### PR TITLE
feat(correlation-id) new "tracker" generator 

### PR DESCRIPTION
**Pull request by WALL-E**
SEE ALSO #1228

### CODE REVIEW

**Comment by Tieske** [comment]
:+1: great! code lgtm 

What's left;

- [ ] update [docs](https://github.com/Mashape/getkong.org)
- [ ] adding [a test](https://github.com/Mashape/kong/blob/next/spec/plugins/correlation-id/access_spec.lua)

File: kong/plugins/correlation-id/handler.lua [hunk 1]
```diff
@@ -11,13 +11,23 @@ local worker_uuid
...
-	end,
+  end,
+  ["tracker"] = function()
+    return string.format("%s-%s-%s-%s-%s-%s",
+      ngx.var.server_addr,
+      ngx.var.server_port,
+      ngx.var.pid,
+      ngx.var.connection, -- connection serial number
+      ngx.var.connection_requests, -- current number of requests made through a connection
+      ngx.now() -- the current time stamp from the nginx cached time.
```
Comment (thibaultCha) [inline kong/plugins/correlation-id/handler.lua]: Could we cache all those global accesses? Something like:

```lua
local fmt = string.format
local now = ngx.now
local pid = ngx.worker.pid -- only need to retrieve it once

-- ...

local var = ngx.var
return fmt(...,
var.server_addr,
var.server_port,
pid,
-- ...
now()
)
```

**Comment by thibaultCha** [comment]
I'm afraid the chances of collision is really high with this generator. `ngx.now` sure provides milliseconds precision but surely that won't be enough to guarantee a level of uniqueness as high as uuids.

**Comment by WALL-E** [comment]
updated doc is done.  [https://github.com/Mashape/getkong.org/pull/236](url)

**Comment by Tieske** [comment]
> I'm afraid the chances of collision is really high with this generator. ngx.now sure provides milliseconds precision but surely that won't be enough to guarantee a level of uniqueness as high as uuids.

yes, crossed my mind as well. But figured it would be easy to add a counter within the millisecond precision if needs be and it actually causes collisions. But maybe better to prevent it any way and add such a counter right away.

**Comment by WALL-E** [comment]
in modern operating system, pid_t is  int type,  and the pid has been increasing(pid%MAX_INT),
at the same time(milliseconds precision), the pid absolutely not repeat.

**Comment by Tieske** [comment]
not sure, but I don't think you should be caching the magic table `ngx` on module level. Sen t you an update.

Not sure I understand your comment (with pid_t) but the question imo is whether we'll __assume__ there will be no collisions, or we code to __make sure__ there are none.

**Comment by timerickson** [comment]
And dont forget subsequent requests may come to different nodes on different hosts who may have varying clocks which could increase the chances of collision.

**Comment by WALL-E** [comment]
Correlation ID = ip-port-pid-connection-connection_requests-timestamp
ip: different hosts
port: different kong node
pid: different process
connection:  serial number
connection_requests: serial number on same connection
timestamp: milliseconds precision

code make sure there are no collisions

File: kong/plugins/correlation-id/handler.lua [hunk 2]
```diff
@@ -10,14 +10,26 @@ local CorrelationIdHandler = BasePlugin:extend()
...
-	end,
+  end,
+  ["tracker"] = function()
+    return fmt("%s-%s-%s-%s-%s-%s",
+      ngx.var.server_addr,
+      ngx.var.server_port,
+      ngx.worker.pid(),
+      ngx.var.connection, -- connection serial number
+      ngx.var.connection_requests, -- current number of requests made through a connection
+      ngx.now() -- the current time stamp from the nginx cached time.
```
Comment (thibaultCha) [inline kong/plugins/correlation-id/handler.lua]: I see you cached `string.format`, but please do follow my other caching suggestions.

**Comment by thibaultCha** [comment]
Right, the combination of `$connection` and `$connection_requests` should avoid any collision indeed. This is still missing a test.

File: kong/plugins/correlation-id/handler.lua (hunk 2 above)
Comment (Tieske) [inline kong/plugins/correlation-id/handler.lua]: I already gave it a quick edit; https://github.com/WALL-E/kong/pull/1

File: spec/plugins/correlation-id/access_spec.lua [hunk 3]
```diff
@@ -79,6 +82,12 @@ describe("Correlation ID Plugin", function()
     end)
   end)
 
+  describe("tracker genetator", function()
```
Comment (Tieske) [inline spec/plugins/correlation-id/access_spec.lua]: typo: generator

**Comment by Tieske** [comment]
@WALL-E can you rebase on `next`?  LGTM, except for the typo. 

@thibaultCha any more comments, or ready to merge? (after rebase)

**Comment by thibaultCha** [comment]
Manually applied the patch with minor modifications to next. Thank you for your contribution!

**Comment by kamoljan** [comment]
Looking forward to next release :)

**Comment by WALL-E** [comment]
I seem to have missed the beautiful moment  :(

**Pull request closed without merging**